File: android/app/src/main/kotlin/com/nomadguy/noctra/lyricromanizer/engines/KoreanEngine.kt
package com.nomadguy.noctra.lyricromanizer.engines

import com.nomadguy.noctra.lyricromanizer.RomanizeEngineContext

/** Korean Hangul romanizer using Jamo decomposition. */
class KoreanEngine {

    private val choseong = listOf(
        "g", "kk", "n", "d", "tt", "r", "m", "b", "pp",
        "s", "ss", "", "j", "jj", "ch", "k", "t", "p", "h"
    )

    private val jungseong = listOf(
        "a", "ae", "ya", "yae", "eo", "e", "yeo", "ye", "o",
        "wa", "wae", "oe", "yo", "u", "wo", "we", "wi", "yu",
        "eu", "ui", "i"
    )

    private val jongseong = listOf(
        "", "k", "k", "ks", "n", "nj", "nh", "t", "l", "lk",
        "lm", "lb", "ls", "lt", "lp", "lh", "m", "p", "ps", "t",
        "t", "ng", "t", "t", "k", "t", "p", "h"
    )

    suspend fun romanize(line: String, context: RomanizeEngineContext): String {
        val buf = StringBuilder()
        for (rune in line.codePoints()) {
            if (rune in 0xAC00..0xD7A3) {
                val code = rune - 0xAC00
                val cho = code / (21 * 28)
                val jung = (code % (21 * 28)) / 28
                val jong = code % 28
                buf.append(choseong[cho])
                buf.append(jungseong[jung])
                buf.append(jongseong[jong])
            } else {
                buf.appendCodePoint(rune)
            }
        }
        return buf.toString()
    }
}

File: lib/services/lyric_romanizer/engines/korean_engine.dart
/// Korean Hangul romanization engine.
///
/// Pure-Dart implementation using Hangul Jamo decomposition
/// with Revised Romanization of Korean tables.

library;
import '../lyric_romanizer_types.dart';

class KoreanRomanizer {
  const KoreanRomanizer();

  static const List<String> _choseong = [
    'g', 'kk', 'n', 'd', 'tt', 'r', 'm', 'b', 'pp',
    's', 'ss', '', 'j', 'jj', 'ch', 'k', 't', 'p', 'h'
  ];

  static const List<String> _jungseong = [
    'a', 'ae', 'ya', 'yae', 'eo', 'e', 'yeo', 'ye', 'o',
    'wa', 'wae', 'oe', 'yo', 'u', 'wo', 'we', 'wi', 'yu',
    'eu', 'ui', 'i'
  ];

  static const List<String> _jongseong = [
    '', 'k', 'k', 'ks', 'n', 'nj', 'nh', 't', 'l', 'lk',
    'lm', 'lb', 'ls', 'lt', 'lp', 'lh', 'm', 'p', 'ps', 't',
    't', 'ng', 't', 't', 'k', 't', 'p', 'h'
  ];

  Future<String> romanize(String line, RomanizeEngineContext context) async {
    final buf = StringBuffer();
    for (final rune in line.runes) {
      if (rune >= 0xAC00 && rune <= 0xD7A3) {
        final code = rune - 0xAC00;
        final cho = code ~/ (21 * 28);
        final jung = (code % (21 * 28)) ~/ 28;
        final jong = code % 28;
        buf.write(_choseong[cho]);
        buf.write(_jungseong[jung]);
        buf.write(_jongseong[jong]);
      } else {
        buf.write(String.fromCharCode(rune));
      }
    }
    return buf.toString();
  }
}
